supermarq-benchmarks/supermarq/qcvv/base_experiment.py

```diff
@@ -0,0 +1,589 @@
+            [sample.circuit for sample in self.samples],
+            target=target,
+            method=method,
+            repetitions=shots,
+            **target_options,
+        )
+        for k, sample in enumerate(self.samples):
+            sample.job = experiment_job[k]
+
+        return experiment_job
+
+    def run_with_simulator(
+        self,
+        simulator: cirq.Sampler | None = None,
+        shots: int = 10_000,
+        overwrite: bool = False,
+    ) -> None:
+        """Use the local simulator to sample the circuits and store the resulting probabilities.
+
+        Args:
+            simulator: A local :class:`~cirq.Sampler` to use. If None then the default
+                :class:`cirq.Simulator` simulator is used. Defaults to None.
+            shots: The number of shots to sample. Defaults to 10,000.
+            overwrite: Whether to force an experiment run even if there is existing data that would
+                be over written in the process. Defaults to False.
+        """
+        if not overwrite:
+            self._run_check()
+
+        if simulator is None:
+            simulator = cirq.Simulator()
+
+        for sample in tqdm(self.samples, desc="Simulating circuits"):
+            result = simulator.run(sample.circuit, repetitions=shots)
+            hist = result.histogram(key=cirq.measurement_key_name(sample.circuit))
+            sample.probabilities = {
+                f"{i:0{self.num_qubits}b}": count / shots for i, count in hist.items()
+            }
+
+    def run_with_callable(
+        self,
+        circuit_eval_func: Callable[[cirq.Circuit], dict[str, float]],
+        overwrite: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        """Evaluates the probabilities for each circuit using a user provided callable function.
+        This function should take a circuit as input and return a dictionary of probabilities for
+        each bitstring (including states with zero probability).
+
+        Args:
+            circuit_eval_func: The custom function to use when evaluating circuit probabilities.
+            overwrite: Whether to force an experiment run even if there is existing data that would
+                be over written in the process. Defaults to False.
+            kwargs: Additional arguments to pass to the custom function.
+
+        Raises:
+            RuntimeError: If the returned probabilities dictionary keys is missing bitstrings.
+            RuntimeError: If the returned probabilities dictionary values do not sum to 1.0.
+        """
+        if not overwrite:
+            self._run_check()
+        for sample in tqdm(self.samples, desc="Running circuits"):
+            probability = circuit_eval_func(sample.circuit, **kwargs)
+            if sorted(probability.keys()) != sorted(
+                f"{i:0{self.num_qubits}b}" for i in range(2**self.num_qubits)
+            ):
+                raise RuntimeError("Returned probabilities are missing bitstrings.")
```

Comment:
I like this way of checking that the bit strings are correct, and rather than enforcing that all probabilities are included I've manually set any missing bitstrings to have zero probability